c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -135,12 +136,35 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                     PointerCast::ArrayToPointer => {}
                     PointerCast::Unsize => {}
                 }
-                self.do_assign_pointer_ids(to_lty.label, from_lty.label)
-                // TODO add other dataflow constraints
+
+                // We would normally do `self.do_assign`,
+                // but we can't do a normal `self.do_equivalence_nested` (inside `self.do_assign`),
+                // so we do the `self.do_assign_pointer_ids` from `self.do_assign`,
+                // and then do the body of `self.do_equivalence_nested`,
+                // but after the extra outer layers are stripped.
+
+                self.do_assign_pointer_ids(to_lty.label, from_lty.label);
+
+                // Each of these ptr casts needs at least the outer layer stripped (e.x. `&`, `*mut`, `fn`),
+                // but then after that, these ptr casts below
+                // needs more layers stripped to reach the types that are equivalent.
+                let strip_from = |lty: LTy<'tcx>| match ptr_cast {
+                    PointerCast::ArrayToPointer => lty.args[0],
+                    PointerCast::Unsize => lty.args[0],
+                    _ => lty,
+                };
+                let strip_to = |lty: LTy<'tcx>| match ptr_cast {
+                    PointerCast::Unsize => lty.args[0],
+                    _ => lty,
+                };
```

Comment:
I think it would be clearer to match on the "from" and "to" `TyKind`s, or else to have a single case per `PointerCast` variant and assert that the types have a particular shape (for both sanity-checking and documentation purposes).  As is, it's hard for me to tell whether or not these rules are correct.  Also, I think matching on `TyKind`s is necessary to distinguish `(x: &[u8; 3]) as &[u8]` from `(x: &[u8; 3]) as &dyn Debug` (both are `PointerCast::Unsize`, but they need different handling here).

Reply:
That's a good point about `dyn` fat ptrs; I hadn't thought of them.  Should I just `error!` on `as &dyn _`s since there shouldn't be any `dyn`s in `lighttpd` or any other `c2rust transpile`d codebase?